File: Kernel Simulator/Misc/Writers/TextWriterColor.vb

'    Kernel Simulator  Copyright (C) 2018-2020  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Console
Imports System.IO
Imports System.Threading

'This module is very important to reduce line numbers when there is color.
Public Module TextWriterColor

    ''' <summary>
    ''' Enumeration for color types
    ''' </summary>
    Public Enum ColTypes As Integer
        Neutral = 1
        Input = 2
        Continuable = 3
        Uncontinuable = 4
        HostName = 5
        UserName = 6
        License = 7
        Gray = 8
        HelpDef = 9
        HelpCmd = 10
        Stage = 11
        Err = 12
    End Enum

    ''' <summary>
    ''' Outputs the text into the terminal prompt, and sets colors as needed.
    ''' </summary>
    ''' <param name="text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Line">Whether to print a new line or not</param>
    ''' <param name="colorType">A type of colors that will be changed.</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub W(ByVal text As Object, ByVal Line As Boolean, ByVal colorType As ColTypes, ByVal ParamArray vars() As Object)

        Dim esc As Char = GetEsc()
        Try
            'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
            If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
                If colorType = ColTypes.Neutral Then
                    Write(esc + "[38;5;" + CStr(neutralTextColor) + "m")
                ElseIf colorType = ColTypes.Input Then
                    Write(esc + "[38;5;" + CStr(inputColor) + "m")
                ElseIf colorType = ColTypes.Continuable Then
                    Write(esc + "[38;5;" + CStr(contKernelErrorColor) + "m")
                ElseIf colorType = ColTypes.Uncontinuable Then
                    Write(esc + "[38;5;" + CStr(uncontKernelErrorColor) + "m")
                ElseIf colorType = ColTypes.HostName Then
                    Write(esc + "[38;5;" + CStr(hostNameShellColor) + "m")
                ElseIf colorType = ColTypes.UserName Then
                    Write(esc + "[38;5;" + CStr(userNameShellColor) + "m")
                ElseIf colorType = ColTypes.License Then
                    Write(esc + "[38;5;" + CStr(licenseColor) + "m")
                ElseIf colorType = ColTypes.Gray Then
                    If backgroundColor = ConsoleColors.DarkYellow Or backgroundColor = ConsoleColors.Yellow Or backgroundColor = ConsoleColors.White Then
                        Write(esc + "[38;5;" + CStr(neutralTextColor) + "m")
                    Else
                        Write(esc + "[38;5;" + CStr(ConsoleColors.Gray) + "m")
                    End If
                ElseIf colorType = ColTypes.HelpDef Then
                    Write(esc + "[38;5;" + CStr(cmdDefColor) + "m")
                ElseIf colorType = ColTypes.HelpCmd Then
                    Write(esc + "[38;5;" + CStr(cmdListColor) + "m")
                ElseIf colorType = ColTypes.Stage Then
                    Write(esc + "[38;5;" + CStr(stageColor) + "m")
                ElseIf colorType = ColTypes.Err Then
                    Write(esc + "[38;5;" + CStr(errorColor) + "m")
                Else
                    Exit Sub
                End If
                Write(esc + "[48;5;" + CStr(backgroundColor) + "m")
            End If

            'Parse variables ({0}, {1}, ...) in the "text" string variable. (Used as a workaround for Linux)
            text = text.ToString.FormatString(vars)

            If Line Then WriteLine(text) Else Write(text)
            If backgroundColor = ConsoleColors.Black Then ResetColor()
        Catch ex As Exception
            WStkTrc(ex)
            KernelError("C", False, 0, DoTranslation("There is a serious error when printing text.", currentLang), ex)
        End Try

    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt slowly with no color support.
    ''' </summary>
    ''' <param name="msg">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Line">Whether to print a new line or not</param>
    ''' <param name="MsEachLetter">Time in milliseconds to delay writing</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteSlowly(ByVal msg As String, ByVal Line As Boolean, ByVal MsEachLetter As Double, ParamArray ByVal vars() As Object)
        'Parse variables ({0}, {1}, ...) in the "text" string variable. (Used as a workaround for Linux)
        msg = msg.FormatString(vars)

        'Write text slowly
        Dim chars As List(Of Char) = msg.ToCharArray.ToList
        For Each ch As Char In chars
            Thread.Sleep(MsEachLetter)
            Write(ch)
        Next
        If Line Then
            WriteLine()
        End If
    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt slowly with color support.
    ''' </summary>
    ''' <param name="msg">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Line">Whether to print a new line or not</param>
    ''' <param name="MsEachLetter">Time in milliseconds to delay writing</param>
    ''' <param name="colorType">A type of colors that will be changed.</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteSlowlyC(ByVal msg As String, ByVal Line As Boolean, ByVal MsEachLetter As Double, ByVal colorType As ColTypes, ParamArray ByVal vars() As Object)
        Dim esc As Char = GetEsc()
        If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
            If colorType = ColTypes.Neutral Then
                Write(esc + "[38;5;" + CStr(neutralTextColor) + "m")
            ElseIf colorType = ColTypes.Input Then
                Write(esc + "[38;5;" + CStr(inputColor) + "m")
            ElseIf colorType = ColTypes.Continuable Then
                Write(esc + "[38;5;" + CStr(contKernelErrorColor) + "m")
            ElseIf colorType = ColTypes.Uncontinuable Then
                Write(esc + "[38;5;" + CStr(uncontKernelErrorColor) + "m")
            ElseIf colorType = ColTypes.HostName Then
                Write(esc + "[38;5;" + CStr(hostNameShellColor) + "m")
            ElseIf colorType = ColTypes.UserName Then
                Write(esc + "[38;5;" + CStr(userNameShellColor) + "m")
            ElseIf colorType = ColTypes.License Then
                Write(esc + "[38;5;" + CStr(licenseColor) + "m")
            ElseIf colorType = ColTypes.Gray Then
                If backgroundColor = ConsoleColors.DarkYellow Or backgroundColor = ConsoleColors.Yellow Or backgroundColor = ConsoleColors.White Then
                    Write(esc + "[38;5;" + CStr(neutralTextColor) + "m")
                Else
                    Write(esc + "[38;5;" + CStr(ConsoleColors.Gray) + "m")
                End If
            ElseIf colorType = ColTypes.HelpDef Then
                Write(esc + "[38;5;" + CStr(cmdDefColor) + "m")
            ElseIf colorType = ColTypes.HelpCmd Then
                Write(esc + "[38;5;" + CStr(cmdListColor) + "m")
            ElseIf colorType = ColTypes.Stage Then
                Write(esc + "[38;5;" + CStr(stageColor) + "m")
            ElseIf colorType = ColTypes.Err Then
                Write(esc + "[38;5;" + CStr(errorColor) + "m")
            Else
                Exit Sub
            End If
            Write(esc + "[48;5;" + CStr(backgroundColor) + "m")
        End If

        'Parse variables ({0}, {1}, ...) in the "text" string variable. (Used as a workaround for Linux)
        msg = msg.FormatString(vars)

        'Write text slowly
        Dim chars As List(Of Char) = msg.ToCharArray.ToList
        For Each ch As Char In chars
            Thread.Sleep(MsEachLetter)
            Write(ch)
        Next
        If Line Then
            WriteLine()
        End If
        If backgroundColor = ConsoleColors.Black Then ResetColor()
    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt with location support, and sets colors as needed.
    ''' </summary>
    ''' <param name="msg">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Left">Column number in console</param>
    ''' <param name="Top">Row number in console</param>
    ''' <param name="colorType">A type of colors that will be changed.</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteWhere(ByVal msg As String, ByVal Left As Integer, ByVal Top As Integer, ByVal colorType As ColTypes, ByVal ParamArray vars() As Object)
        Dim esc As Char = GetEsc()
        If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
            If colorType = ColTypes.Neutral Then
                Write(esc + "[38;5;" + CStr(neutralTextColor) + "m")
            ElseIf colorType = ColTypes.Input Then
                Write(esc + "[38;5;" + CStr(inputColor) + "m")
            ElseIf colorType = ColTypes.Continuable Then
                Write(esc + "[38;5;" + CStr(contKernelErrorColor) + "m")
            ElseIf colorType = ColTypes.Uncontinuable Then
                Write(esc + "[38;5;" + CStr(uncontKernelErrorColor) + "m")
            ElseIf colorType = ColTypes.HostName Then
                Write(esc + "[38;5;" + CStr(hostNameShellColor) + "m")
            ElseIf colorType = ColTypes.UserName Then
                Write(esc + "[38;5;" + CStr(userNameShellColor) + "m")
            ElseIf colorType = ColTypes.License Then
                Write(esc + "[38;5;" + CStr(licenseColor) + "m")
            ElseIf colorType = ColTypes.Gray Then
                If backgroundColor = ConsoleColors.DarkYellow Or backgroundColor = ConsoleColors.Yellow Or backgroundColor = ConsoleColors.White Then
                    Write(esc + "[38;5;" + CStr(neutralTextColor) + "m")
                Else
                    Write(esc + "[38;5;" + CStr(ConsoleColors.Gray) + "m")
                End If
            ElseIf colorType = ColTypes.HelpDef Then
                Write(esc + "[38;5;" + CStr(cmdDefColor) + "m")
            ElseIf colorType = ColTypes.HelpCmd Then
                Write(esc + "[38;5;" + CStr(cmdListColor) + "m")
            ElseIf colorType = ColTypes.Stage Then
                Write(esc + "[38;5;" + CStr(stageColor) + "m")
            ElseIf colorType = ColTypes.Err Then
                Write(esc + "[38;5;" + CStr(errorColor) + "m")
            Else
                Exit Sub
            End If
            Write(esc + "[48;5;" + CStr(backgroundColor) + "m")
        End If

        'Format the message as necessary
        If Not vars.Length = 0 Then msg = String.Format(msg, vars)

        'Write text in another place. By the way, we check the text for newlines and console width excess
        Dim OldLeft As Integer = CursorLeft
        Dim OldTop As Integer = CursorTop
        Dim Paragraphs() As String = msg.SplitNewLines
        SetCursorPosition(Left, Top)
        For MessageParagraphIndex As Integer = 0 To Paragraphs.Length - 1
            'We can now check to see if we're writing a letter past the console window width
            Dim MessageParagraph As String = Paragraphs(MessageParagraphIndex)
            For Each ParagraphChar As Char In MessageParagraph
                If CursorLeft = WindowWidth - 1 Then
                    CursorTop += 1
                    CursorLeft = Left
                End If
                Write(ParagraphChar)
            Next

            'We're starting with the new paragraph, so we increase the CursorTop value by 1.
            If Not MessageParagraphIndex = Paragraphs.Length - 1 Then
                CursorTop += 1
                CursorLeft = Left
            End If
        Next
        SetCursorPosition(OldLeft, OldTop)
        If backgroundColor = ConsoleColors.Black Then ResetColor()
    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt with custom color support.
    ''' </summary>
    ''' <param name="text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Line">Whether to print a new line or not</param>
    ''' <param name="color">A color that will be changed to.</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteC(ByVal text As Object, ByVal Line As Boolean, ByVal color As ConsoleColors, ByVal ParamArray vars() As Object)

        Dim esc As Char = GetEsc()
        Try
            'Try to write to console
            If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
                Write(esc + "[38;5;" + CStr(color) + "m")
                Write(esc + "[48;5;" + CStr(backgroundColor) + "m")
            End If

            'Parse variables ({0}, {1}, ...) in the "text" string variable. (Used as a workaround for Linux)
            text = text.ToString.FormatString(vars)

            If Line Then WriteLine(text) Else Write(text)
            If backgroundColor = ConsoleColors.Black Then ResetColor()
        Catch ex As Exception
            WStkTrc(ex)
            KernelError("C", False, 0, DoTranslation("There is a serious error when printing text.", currentLang), ex)
        End Try

    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt, and sets the true colors as needed.
    ''' </summary>
    ''' <param name="text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Line">Whether to print a new line or not</param>
    ''' <param name="ColorRGBFG">Foreground color RGB storage</param>
    ''' <param name="ColorRGBBG">Background color RGB storage</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteTrueColor(ByVal text As Object, ByVal Line As Boolean, ByVal ColorRGBFG As RGB, ByVal ColorRGBBG As RGB, ByVal ParamArray vars() As Object)

        Dim esc As Char = GetEsc()
        Try
            'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
            If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
                Write(esc + "[38;2;" + ColorRGBFG.ToString + "m")
                Write(esc + "[48;2;" + ColorRGBBG.ToString + "m")
            End If

            'Parse variables ({0}, {1}, ...) in the "text" string variable. (Used as a workaround for Linux)
            text = text.ToString.FormatString(vars)

            If Line Then WriteLine(text) Else Write(text)
            If backgroundColor = ConsoleColors.Black Then ResetColor()
        Catch ex As Exception
            WStkTrc(ex)
            KernelError("C", False, 0, DoTranslation("There is a serious error when printing text.", currentLang), ex)
        End Try

    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt, and sets the true colors as needed.
    ''' </summary>
    ''' <param name="text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Line">Whether to print a new line or not</param>
    ''' <param name="ColorRGB">Color RGB storage</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteTrueColor(ByVal text As Object, ByVal Line As Boolean, ByVal ColorRGB As RGB, ByVal ParamArray vars() As Object)

        Dim esc As Char = GetEsc()
        Try
            'Check if default console output equals the new console output text writer. If it does, write in color, else, suppress the colors.
            If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
                Write(esc + "[38;2;" + ColorRGB.ToString + "m")
                Write(esc + "[48;5;" + CStr(backgroundColor) + "m")
            End If

            'Parse variables ({0}, {1}, ...) in the "text" string variable. (Used as a workaround for Linux)
            text = text.ToString.FormatString(vars)

            If Line Then WriteLine(text) Else Write(text)
            If backgroundColor = ConsoleColors.Black Then ResetColor()
        Catch ex As Exception
            WStkTrc(ex)
            KernelError("C", False, 0, DoTranslation("There is a serious error when printing text.", currentLang), ex)
        End Try

    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt with location support, and sets true colors as needed.
    ''' </summary>
    ''' <param name="msg">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Left">Column number in console</param>
    ''' <param name="Top">Row number in console</param>
    ''' <param name="ColorRGBFG">Foreground color RGB storage</param>
    ''' <param name="ColorRGBBG">Background color RGB storage</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteWhereTrueColor(ByVal msg As String, ByVal Left As Integer, ByVal Top As Integer, ByVal ColorRGBFG As RGB, ByVal ColorRGBBG As RGB, ByVal ParamArray vars() As Object)
        Dim esc As Char = GetEsc()
        If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
            Write(esc + "[38;2;" + ColorRGBFG.ToString + "m")
            Write(esc + "[48;2;" + ColorRGBBG.ToString + "m")
        End If

        'Format the message as necessary
        If Not vars.Length = 0 Then msg = String.Format(msg, vars)

        'Write text in another place. By the way, we check the text for newlines and console width excess
        Dim OldLeft As Integer = CursorLeft
        Dim OldTop As Integer = CursorTop
        Dim Paragraphs() As String = msg.SplitNewLines
        SetCursorPosition(Left, Top)
        For MessageParagraphIndex As Integer = 0 To Paragraphs.Length - 1
            'We can now check to see if we're writing a letter past the console window width
            Dim MessageParagraph As String = Paragraphs(MessageParagraphIndex)
            For Each ParagraphChar As Char In MessageParagraph
                If CursorLeft = WindowWidth - 1 Then
                    CursorTop += 1
                    CursorLeft = Left
                End If
                Write(ParagraphChar)
            Next

            'We're starting with the new paragraph, so we increase the CursorTop value by 1.
            If Not MessageParagraphIndex = Paragraphs.Length - 1 Then
                CursorTop += 1
                CursorLeft = Left
            End If
        Next
        SetCursorPosition(OldLeft, OldTop)
        If backgroundColor = ConsoleColors.Black Then ResetColor()
    End Sub

    ''' <summary>
    ''' Outputs the text into the terminal prompt with location support, and sets true colors as needed.
    ''' </summary>
    ''' <param name="msg">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
    ''' <param name="Left">Column number in console</param>
    ''' <param name="Top">Row number in console</param>
    ''' <param name="ColorRGB">Color RGB storage</param>
    ''' <param name="vars">Endless amounts of any variables that is separated by commas.</param>
    Public Sub WriteWhereTrueColor(ByVal msg As String, ByVal Left As Integer, ByVal Top As Integer, ByVal ColorRGB As RGB, ByVal ParamArray vars() As Object)
        Dim esc As Char = GetEsc()
        If IsNothing(DefConsoleOut) Or Equals(DefConsoleOut, Out) Then
            Write(esc + "[38;2;" + ColorRGB.ToString + "m")
            Write(esc + "[48;5;" + CStr(backgroundColor) + "m")
        End If

        'Format the message as necessary
        If Not vars.Length = 0 Then msg = String.Format(msg, vars)

        'Write text in another place. By the way, we check the text for newlines and console width excess
        Dim OldLeft As Integer = CursorLeft
        Dim OldTop As Integer = CursorTop
        Dim Paragraphs() As String = msg.SplitNewLines
        SetCursorPosition(Left, Top)
        For MessageParagraphIndex As Integer = 0 To Paragraphs.Length - 1
            'We can now check to see if we're writing a letter past the console window width
            Dim MessageParagraph As String = Paragraphs(MessageParagraphIndex)
            For Each ParagraphChar As Char In MessageParagraph
                If CursorLeft = WindowWidth - 1 Then
                    CursorTop += 1
                    CursorLeft = Left
                End If
                Write(ParagraphChar)
            Next

            'We're starting with the new paragraph, so we increase the CursorTop value by 1.
            If Not MessageParagraphIndex = Paragraphs.Length - 1 Then
                CursorTop += 1
                CursorLeft = Left
            End If
        Next
        SetCursorPosition(OldLeft, OldTop)
        If backgroundColor = ConsoleColors.Black Then ResetColor()
    End Sub

End Module

File: public/Nitrocid/ConsoleBase/Writers/ConsoleWriters/TextWriterColor.cs
//
// Nitrocid KS  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Nitrocid KS
//
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Threading;
using Terminaux.Colors;
using Nitrocid.Kernel.Debugging;
using Nitrocid.Languages;
using Nitrocid.ConsoleBase.Colors;

namespace Nitrocid.ConsoleBase.Writers.ConsoleWriters
{
    /// <summary>
    /// Console text writer with color support
    /// </summary>
    public static class TextWriterColor
    {

        internal static object WriteLock = new();

        /// <summary>
        /// Outputs the new line into the terminal prompt, and sets colors as needed.
        /// </summary>
        public static void Write()
        {
            lock (WriteLock)
            {
                ConsoleWrapper.WriteLine();
            }
        }

        /// <summary>
        /// Outputs the text into the terminal prompt plainly.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WritePlain(string Text, params object[] vars) =>
            WritePlain(Text, true, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt plainly.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WritePlain(string Text, bool Line, params object[] vars)
        {
            lock (WriteLock)
            {
                try
                {
                    // Actually write
                    if (Line)
                    {
                        if (vars.Length > 0)
                        {
                            ConsoleWrapper.WriteLine(Text, vars);
                        }
                        else
                        {
                            ConsoleWrapper.WriteLine(Text);
                        }
                    }
                    else if (vars.Length > 0)
                    {
                        ConsoleWrapper.Write(Text, vars);
                    }
                    else
                    {
                        ConsoleWrapper.Write(Text);
                    }
                }
                catch (Exception ex) when (ex.GetType().Name != nameof(ThreadInterruptedException))
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    DebugWriter.WriteDebug(DebugLevel.E, Translate.DoTranslation("There is a serious error when printing text.") + " {0}", ex.Message);
                }
            }
        }

        /// <summary>
        /// Outputs the text into the terminal prompt.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void Write(string Text, params object[] vars) =>
            Write(Text, true, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void Write(string Text, bool Line, params object[] vars) =>
            Write(Text, Line, false, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="Highlight">Highlight the text written</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void Write(string Text, bool Line, bool Highlight, params object[] vars) =>
            WriteColor(Text, Line, Highlight, KernelColorTools.GetColor(KernelColorType.NeutralText), vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="color">A color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColor(string Text, ConsoleColors color, params object[] vars) =>
            WriteColor(Text, true, false, color, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="color">A color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColor(string Text, bool Line, ConsoleColors color, params object[] vars) =>
            WriteColor(Text, Line, false, color, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="Highlight">Highlight the text written</param>
        /// <param name="color">A color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColor(string Text, bool Line, bool Highlight, ConsoleColors color, params object[] vars)
        {
            lock (WriteLock)
            {
                try
                {
                    // Try to write to console
                    KernelColorTools.SetConsoleColor(new Color(color), Highlight);
                    KernelColorTools.SetConsoleColor(KernelColorType.Background, !Highlight, false);

                    // Write the text to console
                    if (Highlight)
                    {
                        WritePlain(Text, false, vars);
                        KernelColorTools.SetConsoleColor(new Color(color));
                        KernelColorTools.SetConsoleColor(KernelColorType.Background, true);
                        WritePlain("", Line);
                    }
                    else
                    {
                        WritePlain(Text, Line, vars);
                    }
                }
                catch (Exception ex) when (ex.GetType().Name != nameof(ThreadInterruptedException))
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    DebugWriter.WriteDebug(DebugLevel.E, Translate.DoTranslation("There is a serious error when printing text.") + " {0}", ex.Message);
                }
            }
        }

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColorBack(string Text, ConsoleColors ForegroundColor, ConsoleColors BackgroundColor, params object[] vars) =>
            WriteColorBack(Text, true, false, ForegroundColor, BackgroundColor, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColorBack(string Text, bool Line, ConsoleColors ForegroundColor, ConsoleColors BackgroundColor, params object[] vars) =>
            WriteColorBack(Text, Line, false, ForegroundColor, BackgroundColor, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="Highlight">Highlight the text written</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColorBack(string Text, bool Line, bool Highlight, ConsoleColors ForegroundColor, ConsoleColors BackgroundColor, params object[] vars)
        {
            lock (WriteLock)
            {
                try
                {
                    // Try to write to console
                    KernelColorTools.SetConsoleColor(new Color(ForegroundColor), Highlight);
                    KernelColorTools.SetConsoleColor(new Color(BackgroundColor), !Highlight);

                    // Write the text to console
                    if (Highlight)
                    {
                        WritePlain(Text, false, vars);
                        KernelColorTools.SetConsoleColor(new Color(ForegroundColor));
                        KernelColorTools.SetConsoleColor(new Color(BackgroundColor), true);
                        WritePlain("", Line);
                    }
                    else
                    {
                        WritePlain(Text, Line, vars);
                    }
                }
                catch (Exception ex) when (ex.GetType().Name != nameof(ThreadInterruptedException))
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    DebugWriter.WriteDebug(DebugLevel.E, Translate.DoTranslation("There is a serious error when printing text.") + " {0}", ex.Message);
                }
            }
        }

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="color">A color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColor(string Text, Color color, params object[] vars) =>
            WriteColor(Text, true, false, color, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="color">A color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColor(string Text, bool Line, Color color, params object[] vars) =>
            WriteColor(Text, Line, false, color, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="Highlight">Highlight the text written</param>
        /// <param name="color">A color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColor(string Text, bool Line, bool Highlight, Color color, params object[] vars)
        {
            lock (WriteLock)
            {
                try
                {
                    // Try to write to console
                    KernelColorTools.SetConsoleColor(color, Highlight);
                    KernelColorTools.SetConsoleColor(KernelColorType.Background, !Highlight, false);

                    // Write the text to console
                    if (Highlight)
                    {
                        WritePlain(Text, false, vars);
                        KernelColorTools.SetConsoleColor(color);
                        KernelColorTools.SetConsoleColor(KernelColorType.Background, true);
                        WritePlain("", Line);
                    }
                    else
                    {
                        WritePlain(Text, Line, vars);
                    }
                }
                catch (Exception ex) when (ex.GetType().Name != nameof(ThreadInterruptedException))
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    DebugWriter.WriteDebug(DebugLevel.E, Translate.DoTranslation("There is a serious error when printing text.") + " {0}", ex.Message);
                }
            }
        }

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColorBack(string Text, Color ForegroundColor, Color BackgroundColor, params object[] vars) =>
            WriteColorBack(Text, true, false, ForegroundColor, BackgroundColor, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColorBack(string Text, bool Line, Color ForegroundColor, Color BackgroundColor, params object[] vars) =>
            WriteColorBack(Text, Line, false, ForegroundColor, BackgroundColor, vars);

        /// <summary>
        /// Outputs the text into the terminal prompt with custom color support.
        /// </summary>
        /// <param name="Text">A sentence that will be written to the terminal prompt. Supports {0}, {1}, ...</param>
        /// <param name="Line">Whether to print a new line or not</param>
        /// <param name="Highlight">Highlight the text written</param>
        /// <param name="ForegroundColor">A foreground color that will be changed to.</param>
        /// <param name="BackgroundColor">A background color that will be changed to.</param>
        /// <param name="vars">Variables to format the message before it's written.</param>
        public static void WriteColorBack(string Text, bool Line, bool Highlight, Color ForegroundColor, Color BackgroundColor, params object[] vars)
        {
            lock (WriteLock)
            {
                try
                {
                    // Try to write to console
                    KernelColorTools.SetConsoleColor(ForegroundColor, Highlight);
                    KernelColorTools.SetConsoleColor(BackgroundColor, !Highlight);

                    // Write the text to console
                    if (Highlight)
                    {
                        WritePlain(Text, false, vars);
                        KernelColorTools.SetConsoleColor(ForegroundColor);
                        KernelColorTools.SetConsoleColor(BackgroundColor, true);
                        WritePlain("", Line);
                    }
                    else
                    {
                        WritePlain(Text, Line, vars);
                    }
                }
                catch (Exception ex) when (ex.GetType().Name != nameof(ThreadInterruptedException))
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    DebugWriter.WriteDebug(DebugLevel.E, Translate.DoTranslation("There is a serious error when printing text.") + " {0}", ex.Message);
                }
            }
        }

    }
}
